### ISSUE TITLE
Missing AZEL column?

### ISSUE THREAD

**Issue by mknapp55** [lofar-astron/prefactor#134]
I'm trying to run the Initial-Subtract-Fast parset on data prepared by the calibration and target pipelines.  I'm getting the following error:

`2017-05-16 03:28:19 ERROR   node.ip-100-64-109-221.python_plugin: TableProxy::getColumn: column AZEL1 does not exist`

I previously ran the 'normal' Initial-Subtract on this data without issues (other than the very long runtime).

**Comment by darafferty** [comment]
There error must be occurring in line 69 or 71 of https://github.com/lofar-astron/prefactor/blob/master/scripts/InitSubtract_sort_and_compute.py (which is strange as this script is also used in the normal Intitial-Subtract pipeline). You should be able to get more info about the error if you comment out the try/except block in this script that begins on line 60 (so that the RuntimeError is not caught).

**Comment by mknapp55** [comment]
@darafferty Here's the result from commenting out the try/except:
`2017-05-16 18:18:22 DEBUG   genericpipeline.executable_args: InitSubtract_sort_and_compute.py: Putting files into bands.
2017-05-16 18:18:22 DEBUG   genericpipeline.executable_args: Results for job 0 submitted by ('127.0.0.1', 41586)
2017-05-16 18:18:22 DEBUG   genericpipeline.executable_args: InitSubtract_sort_and_compute.py: Working on Band: 169
2017-05-16 18:18:22 ERROR   node.ip-100-64-109-221.python_plugin: Table DataManager error: Data Manager class  is not registered
  Check (DY)LD_LIBRARY_PATH matches the libraries used during the build of 
2017-05-16 18:18:22 INFO    node.ip-100-64-109-221.python_plugin: Total time 1.1986s; user time: 0.4360s; system time: 0.8440s
2017-05-16 18:18:22 DEBUG   node.ip-100-64-109-221.python_plugin: Start time was 1494958701.6719s; end time was 1494958702.8706s
`

**Comment by AHorneffer** [comment]
@mknapp55 AFAIK the `pt.addDerivedMSCal()` call adds virtual columns to the MS, which are computed on the fly, so you don't see them with casabrowser.

It looks like there is something screwed up with your installation of casacore (or python-casacore). I believe this could happen if you have the libraries of a standalone casa(py) in your `LD_LIBRARY_PATH`.

**Comment by mknapp55** [comment]
@AHorneffer There's nothing in my `LD_LIBRARY_PATH` - the environmental variable doesn't exist.  Does prefactor need this env variable, and if so what should it point to?  I would check what it is on CEP3, but it's down for maintenance.

I'm using Kern for casacore, pyrap, lofar, etc.  It's worked fine in the past.  I'm using the PiLL pipeline in this environment with no issues.  Any ideas about what conflicts/bad paths I should look for?

**Comment by AHorneffer** [comment]
If you run the pipeline with debug output, then it will print the paths that are actually used to the logfile. Can you put the full logfile somewhere where we can have a look at it?

See also: [FAQ: My pipeline crashes and I cannot find the problem here](https://github.com/lofar-astron/prefactor/wiki/Documentation%3A-Faq#my-pipeline-crashes-and-i-cannot-find-the-problem-here)

**Comment by mknapp55** [comment]
See attached.
[pipeline.txt](https://github.com/lofar-astron/prefactor/files/1010778/pipeline.txt)

**Comment by AHorneffer** [comment]
O.K. `LD_LIBRARY_PATH` indeed isn't set.

Can you check which (shared-)libraries `pyrap.tables` (Or `casacore.tables`, that's the same) loads?

You can see that by running `ldd` on the `_tables.so` library of the "pyrap" (or "python_casacore") package. On my machine here I can do that with:
```
ldd /opt/soft/lofar-stuff/lib/python2.7/site-packages/python_casacore-2.0.0-py2.7-linux-x86_64.egg/casacore/tables/_tables.so
```

P.S. As David already mentioned: the `InitSubtract_sort_and_compute.py` is the same in the `Initial-Subtract.parset` and the `Initial-Subtract-Fast.parset`. So something must have changed during the two runs. And mixing casacore libraries from a LOFAR software installation and from casapy is known to cause problems.

[...]

Well, another possibility: do you use dysco compression on the MSs? If so, is the dysco library found when you run the pipeline?

**Comment by mknapp55** [comment]
I apparently don't have `_tables.so` anywhere on my system.  Here's the result for libcasa_tables.so:
```
ubuntu@ip-100-64-109-221:~$ ldd /usr/lib/libcasa_tables.so
	linux-vdso.so.1 =>  (0x00007fffcf622000)
	libcasa_casa.so.1 => /usr/lib/libcasa_casa.so.1 (0x00007f2d33bcc000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f2d33995000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f2d3368b000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f2d33309000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f2d330f3000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f2d32d29000)
	libhdf5.so.7 => /usr/lib/x86_64-linux-gnu/libhdf5.so.7 (0x00007f2d3288d000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f2d32689000)
	/lib64/ld-linux-x86-64.so.2 (0x0000561cc5217000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f2d3246e000)
```
I don't know why I don't have _tables.so - I'm guessing it's something to do with the Kern installation?

I do have Dysco installed, but I don't think I used it in any of the previous steps...

**Comment by darafferty** [comment]
Could the problem be that the `LD_LIBRARY_PATH` is not set? I have the path to the casacore libraries in mine, but I don't know for sure that it's needed.

**Comment by mknapp55** [comment]
What should that path be set to?  Normally, I would check what it is on
CEP3, but that isn't an option right now.

**Comment by mknapp55** [comment]
No, those are not set explicitly.  I modified the pipeline.cfg that came with my lofar installation (via Kern) (`$LOFARROOT/share/pipeline/pipeline.cfg`).  My pipeline.cfg is below.  I use the same base pipeline.cfg with the PiLL pipeline (which uses the generic pipeline) with no issues.

```
[DEFAULT]
lofarroot = /usr
casaroot = 
pyraproot = 
hdf5root = 
wcsroot = 
aoflaggerroot=
pythonpath = /usr/lib/python2.7/dist-packages
runtime_directory = /data/t1data/hba/runtime
recipe_directories = [%(pythonpath)s/lofarpipe/recipes,/data/prefactor]
working_directory = /data/t1data/hba/working
task_files = [%(lofarroot)s/share/pipeline/tasks.cfg]

[layout]
job_directory = %(runtime_directory)s/%(job_name)s

[cluster]
clusterdesc = %(lofarroot)s/share/local.clusterdesc

[deploy]
engine_ppath = %(pythonpath)s:%(pyraproot)s/lib:/opt/cep/pythonlibs/lib/python/site-packages
engine_lpath = %(lofarroot)s/lib:%(casaroot)s/lib:%(pyraproot)s/lib:%(hdf5root)s/lib:%(wcsroot)s/lib

[logging]
log_file = %(runtime_directory)s/%(job_name)s/logs/%(start_time)s/pipeline.log
xml_stat_file = %(runtime_directory)s/%(job_name)s/logs/%(start_time)s/statistics.xml

[feedback]
# Method of providing feedback to LOFAR.
# Valid options:
#    messagebus    Send feedback and status using LCS/MessageBus
#    none          Do NOT send feedback and status
method = none

[remote]
method = local
max_per_node = 2
```

**Comment by AHorneffer** [comment]
@mknapp55 Well, did you try setting `casaroot` and `pyraproot`?

**Comment by mknapp55** [comment]
Same result.  I've attached the log and the top block of pipeline.cfg is below.

```
[DEFAULT]
lofarroot = /usr
casaroot = /usr
pyraproot = /usr/lib/python2.7/dist-packages
hdf5root = 
wcsroot = 
aoflaggerroot=
pythonpath = /usr/lib/python2.7/dist-packages
runtime_directory = /data/t1data/hba/runtime
recipe_directories = [%(pythonpath)s/lofarpipe/recipes,/data/prefactor]
working_directory = /data/t1data/hba/working
task_files = [%(lofarroot)s/share/pipeline/tasks.cfg]
```
[pipeline2.txt](https://github.com/lofar-astron/prefactor/files/1011270/pipeline2.txt)

**Comment by darafferty** [comment]
My `LD_LIBRARY_PATH` has the path to the directory with `libcasa_tables.so` in it (so I guess `/usr/lib` for you?).

**Comment by mknapp55** [comment]
Hmmm...I have the following rather than `libcasa_tables.so`:
```
mknapp@mimas:/usr/lib$ locate libcasa_tables.so
/usr/lib/libcasa_tables.so.2
/usr/lib/libcasa_tables.so.2.0.3
```
@tammojan said this was a common problem with the Kern installation - it makes somelib.so.# rather than somelib.so.  I can try to sym link my `libcasa_tables.so.*` to `libcasa_tables.so`...

**Comment by AHorneffer** [comment]
@mknapp55 Try setting the `LD_LIBRARY_PATH` (to `/usr/lib` I guess) before running the pipeline.

If that doesn't help, then search for the `_tables.so` library from "python_casacore" and run `ldd` on it. (I guess for you it would be in `/usr/lib/python2.7/dist-packages/python_casacore-*.egg/casacore/tables/_tables.so`. If you can't find it, do `import pyrap.tables as pt` in a python shell and then `help(pt)`.)

**Comment by mknapp55** [comment]
@AHorneffer That library (`_tables.so`) doesn't exist on my system (it doesn't turn up using `locate _tables.so`).  This is the closest I could get to the directory you suggested:
```
ubuntu@ip-100-64-109-221:/usr/lib/python2.7/dist-packages/python_casacore-2.1.2.egg-info$ ls
dependency_links.txt  PKG-INFO  requires.txt  top_level.txt
```

Here's the result from `help(pt)`:
```
Help on module pyrap.tables in pyrap:

NAME
    pyrap.tables

FILE
    /usr/lib/python2.7/dist-packages/pyrap/tables.py
```

**Comment by AHorneffer** [comment]
@mknapp55 Well, when you do `import pyrap.tables` in python it sooner or later has to load a compiled library that includes the bindings to `libcasa_tables`. Can you search for that library?

tl;dr I don't believe that you have a working pyrap but don't have a `_tables.so` on your system somewhere.

**Comment by mknapp55** [comment]
@AHorneffer Yes, libcasa_tables.so lives in /usr/lib/
```
ubuntu@ip-100-64-109-221:/usr/lib$ ls libcasa_tables*
libcasa_tables.so  libcasa_tables.so.1  libcasa_tables.so.1.7.0
```

**Comment by AHorneffer** [comment]
@mknapp55 Is that directly loaded from the python scripts?

**Comment by mknapp55** [comment]
@AHorneffer I'm not sure what you mean?

**Comment by AHorneffer** [comment]
@mknapp55 Here I get that:
```
lofard3:horneff> cat /opt/soft/lofar-stuff/lib/python2.7/site-packages/python_casacore-2.0.0-py2.7-linux-x86_64.egg/casacore/tables/_tables.py
def __bootstrap__():
    global __bootstrap__, __loader__, __file__
    import sys, pkg_resources, imp
    __file__ = pkg_resources.resource_filename(__name__, '_tables.so')
    __loader__ = None; del __bootstrap__, __loader__
    imp.load_dynamic(__name__,__file__)
__bootstrap__()
```

That piece of python code tells the python interpreter to load the dynamic library `_tables.so` (which then in turn links the `libcasa_tables.so`), it gets executed then I do `import pyrap.tables` in a python interpreter.

**Comment by mknapp55** [comment]
This directory `python_casacore-2.0.0-py2.7-linux-x86_64.egg` (or similar) does not exist on my system.  Yet, I have no trouble using pyrap in `ipython`.  I can only guess this is a Kern issue, but I don't understand why I can use pyrap, on its own, but something goes wrong in that script.
```
ubuntu@ip-100-64-109-221:/usr/lib$ locate pyrap
/usr/lib/python2.7/dist-packages/pyrap
/usr/lib/python2.7/dist-packages/pyrap/__init__.py
/usr/lib/python2.7/dist-packages/pyrap/__init__.pyc
/usr/lib/python2.7/dist-packages/pyrap/fitting.py
/usr/lib/python2.7/dist-packages/pyrap/fitting.pyc
/usr/lib/python2.7/dist-packages/pyrap/functionals.py
/usr/lib/python2.7/dist-packages/pyrap/functionals.pyc
/usr/lib/python2.7/dist-packages/pyrap/images
/usr/lib/python2.7/dist-packages/pyrap/images.py
/usr/lib/python2.7/dist-packages/pyrap/images.pyc
/usr/lib/python2.7/dist-packages/pyrap/measures.py
/usr/lib/python2.7/dist-packages/pyrap/measures.pyc
/usr/lib/python2.7/dist-packages/pyrap/quanta.py
/usr/lib/python2.7/dist-packages/pyrap/quanta.pyc
/usr/lib/python2.7/dist-packages/pyrap/tables.py
/usr/lib/python2.7/dist-packages/pyrap/tables.pyc
/usr/lib/python2.7/dist-packages/pyrap/util.py
/usr/lib/python2.7/dist-packages/pyrap/util.pyc
/usr/lib/python2.7/dist-packages/pyrap/images/__init__.py
/usr/lib/python2.7/dist-packages/pyrap/images/__init__.pyc
/usr/lib/python2.7/dist-packages/pyrap/images/coordinates.py
/usr/lib/python2.7/dist-packages/pyrap/images/coordinates.pyc
/usr/lib/python2.7/dist-packages/pyrap/images/image.py
/usr/lib/python2.7/dist-packages/pyrap/images/image.pyc
```

**Comment by AHorneffer** [comment]
**bleep**ing Kern!

Do you have something like: `/usr/lib/python2.7/dist-packages/casacore/tables/_tables.x86_64-linux-gnu.so` or similar? (anything that starts with `_tables` and looks like a dynamically loaded library.) Can you run `ldd` on it?

**Comment by mknapp55** [comment]
@AHorneffer I do!
```
ubuntu@ip-100-64-109-221:/usr/lib/python2.7/dist-packages/casacore/tables$ ldd _tables.x86_64-linux-gnu.so 
	linux-vdso.so.1 =>  (0x00007ffe97f33000)
	libcasa_tables.so.2 => /usr/lib/x86_64-linux-gnu/libcasa_tables.so.2 (0x00007f988d1bc000)
	libboost_python-py27.so.1.58.0 => /usr/lib/x86_64-linux-gnu/libboost_python-py27.so.1.58.0 (0x00007f988cf70000)
	libcasa_python.so.2 => /usr/lib/x86_64-linux-gnu/libcasa_python.so.2 (0x00007f988cd36000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f988c9b4000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f988c79e000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f988c3d4000)
	libcasa_casa.so.2 => /usr/lib/x86_64-linux-gnu/libcasa_casa.so.2 (0x00007f988be3a000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f988bb31000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f988b913000)
	libpython2.7.so.1.0 => /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0 (0x00007f988b385000)
	/lib64/ld-linux-x86-64.so.2 (0x000055adfc7e9000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f988b180000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f988af66000)
	libutil.so.1 => /lib/x86_64-linux-gnu/libutil.so.1 (0x00007f988ad63000)
```

**Comment by darafferty** [comment]
I looks like the paths for the libraries that casacore.tables uses are in `/usr/lib/x86_64-linux-gnu` instead of `/usr/lib`. Maybe try adding `/usr/lib/x86_64-linux-gnu` to your `LD_LIBRARY_PATH`?

**Comment by mknapp55** [comment]
@darafferty I added  `/usr/lib/x86_64-linux-gnu` to `$LD_LIBRARY_PATH`....and got the same error.  I deleted and redownloaded prefactor just in case I'd broken something, and that didn't help either.  The paths that the script found are:

```
 2017-05-19 23:07:12 DEBUG   node.ip-100-64-109-221.python_plugin: environment       = {'OMP_NUM_THREADS': '8', 'PATH': '/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin', 'LOFARROOT': '/usr', 'LD_LIBRARY_PATH': '/usr/lib/x86_64-linux-gnu:/usr/lib', 'PYTHONPATH': '/usr/lib/python2.7/dist-packages:/data/scripts/'}
```
The error is back to the `AZEL1` error since my fresh download of prefactor doesn't have the try/catch commented out.
```
2017-05-19 23:07:13 DEBUG   genericpipeline.executable_args: InitSubtract_sort_and_compute.py: Putting files into bands.
2017-05-19 23:07:13 DEBUG   genericpipeline.executable_args: InitSubtract_sort_and_compute.py: Working on Band: 169
2017-05-19 23:07:13 DEBUG   genericpipeline.executable_args: Results for job 0 submitted by ('127.0.0.1', 43802)
2017-05-19 23:07:13 ERROR   node.ip-100-64-109-221.python_plugin: TableProxy::getColumn: column AZEL1 does not exist
2017-05-19 23:07:13 INFO    node.ip-100-64-109-221.python_plugin: Total time 1.4854s; user time: 0.2320s; system time: 0.0320s
2017-05-19 23:07:13 DEBUG   node.ip-100-64-109-221.python_plugin: Start time was 1495235232.2086s; end time was 1495235233.6942s
2017-05-19 23:07:14 DEBUG   genericpipeline.executable_args: 
2017-05-19 23:07:14 WARNING genericpipeline.executable_args: 
2017-05-19 23:07:14 INFO    genericpipeline.executable_args: Subprocess completed with exit status 1: /bin/sh -c python /usr/lib/python2.7/dist-packages/lofarpipe/recipes/nodes/python_plugin.py 0 127.0.1.1 44923
```
I'm wondering if perhaps the problem is that these MSs went part way through an Init-Subtract run?  Maybe I should rerun the target pipeline to get 'fresh' MSs for Init-Subtract??

**Comment by AHorneffer** [comment]
It still look to me like a problem with your casacore / python-casacore installation.

Can you try to run the relevant pieces of code by hand? E.g. something like this:
```
import pyrap.tables as pt
my_ms = "/path/to/my_ms_file.ms"
pt.addDerivedMSCal(my_ms)
tab = pt.table(self.files[MS_id], ack=False)
global_el_values = tab.getcol('AZEL1', rowincr=10000)[:, 1]
tab.close()
pt.removeDerivedMSCal(my_ms)
```
If the  `pt.addDerivedMSCal()` call raises an exception that reads like it is complaining the the columns already exist, then please continue. 
If it fails on your data, can you run it on a "clean" MS (an MS where you a re sure that corrupted data in the MS is not a problem).

**Comment by mknapp55** [comment]
I get the same Runtime Error as in the logs.  It's complaining about `LD_LIBRARY_PATH`, but I've already tried setting/changing that and it hasn't made a difference.  I set it to `LD_LIBRARY_PATH=/usr/lib:/usr/lib/x86_64-linux-gnu` and got the same results.  The `AZEL1` column still appears to be missing as well.  I got the same results on an LBA MS straight off the LTA.
```
In [1]: import pyrap.tables as pt

In [2]: my_ms = '/data/t1data/hba/results_L523928_calpost/L523928_SB000_uv.dppp_1287E19ABt_159MHz.pre-cal.ms/'

In [3]: pt.addDerivedMSCal(my_ms)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-3-d266bf2f8376> in <module>()
----> 1 pt.addDerivedMSCal(my_ms)

/usr/lib/python2.7/dist-packages/casacore/tables/msutil.pyc in addDerivedMSCal(msname)
    170     # Add all columns using DerivedMSCal as data manager.
    171     dminfo = {"TYPE": "DerivedMSCal", "NAME": "", "SPEC": {}}
--> 172     t.addcols (maketabdesc(descs), dminfo)
    173     # Flush the table to make sure it is written.
    174     t.flush()

/usr/lib/python2.7/dist-packages/casacore/tables/table.pyc in addcols(self, desc, dminfo, addtoparent)
   1137                 cd = casacore.tables.tableutil.makecoldesc (desc['name'], desc)
   1138                 tdesc = casacore.tables.tableutil.maketabdesc(cd)
-> 1139         self._addcols (tdesc, dminfo, addtoparent)
   1140         self._makerow()
   1141 

RuntimeError: Table DataManager error: Data Manager class  is not registered
  Check (DY)LD_LIBRARY_PATH matches the libraries used during the build of

In [4]: tab = pt.table(self.files[MS_id], ack=False)
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-4-29de700a242c> in <module>()
----> 1 tab = pt.table(self.files[MS_id], ack=False)

NameError: name 'self' is not defined

In [5]: tab = pt.table(my_ms, ack=False)

In [6]: global_el_values = tab.getcol('AZEL1', rowincr=10000)[:, 1]
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-6-ffd03a496cfd> in <module>()
----> 1 global_el_values = tab.getcol('AZEL1', rowincr=10000)[:, 1]

/usr/lib/python2.7/dist-packages/casacore/tables/table.pyc in getcol(self, columnname, startrow, nrow, rowincr)
    949 #                i = inx*
    950 #        except:
--> 951         return self._getcol (columnname, startrow, nrow, rowincr)
    952 
    953     def getcolnp (self, columnname, nparray, startrow=0, nrow=-1, rowincr=1):

RuntimeError: TableProxy::getColumn: column AZEL1 does not exist

In [7]: tab.close()

In [8]: pt.removeDerivedMSCal(my_ms)

In [9]: 
```

**Comment by AHorneffer** [comment]
Still looks like a problem with the software installation. (Somewhere at the boundary between python code and the compiled C(++) code some module / class isn't found.)
I assume you are using python 2.7 and that this is the first python-interpreter in your path.

I have no more idea how to fix this, but it doesn't look like a bug in prefactor to me.
(Well, except to re-install the software.)

P.S. That below was a copy&paste error of mine. It should be `my_ms` instead of `self.files[MS_id]`.
> In [4]: tab = pt.table(self.files[MS_id], ack=False)
> ---------------------------------------------------------------------------
> NameError                                 Traceback (most recent call last) 
> <ipython-input-4-29de700a242c> in <module>()
> ----> 1 tab = pt.table(self.files[MS_id], ack=False)
> 
> NameError: name 'self' is not defined

**Comment by mknapp55** [comment]
I'll open an issue with Kern.

**Comment by sabourke** [comment]
Just encountered this myself. In case it's still an issue for you, you need the libcasa-derivedmscal2 package installed.
